### PR TITLE
Shorten uid in BlogPost to make it better suited for short urls, closes #227

### LINKED ISSUE #227: Migrate to shorter uids that could be used as a part of shortened links

### PR DESCRIPTION
**Pull request by tdziurko**

### CODE REVIEW

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/data_migration/BlogPostUidShortener.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        this.blogPostRepository = blogPostRepository;
+    }
+
+    int shortenUUids(int batchSize) {
+        List<BlogPost> postsToShortenUid = blogPostRepository
+                .findPostsWithUidLongerThan(BlogPost.UID_LENGTH, new PageRequest(0, batchSize));
+
+        postsToShortenUid.stream().forEach(post -> {
+            post.setUid(RandomStringUtils.randomAlphanumeric(BlogPost.UID_LENGTH));
+            blogPostRepository.save(post);
```
Comment (mszarlinski) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/data_migration/BlogPostUidShortener.java]: If the method was annotated with @Transactional, updating posts would be done automatically by Hibernate, so this line could be skipped.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/data_migration/BlogPostUidShortener.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import java.util.List;
+
+@Component
+class BlogPostUidShortener {
+
+    private final BlogPostRepository blogPostRepository;
+
+    @Autowired
+    public BlogPostUidShortener(BlogPostRepository blogPostRepository) {
```
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/data_migration/BlogPostUidShortener.java]: This handcrafted constructor can be replaced with class level annotation [`@RequiredArgsConstructor(onConstructor = @__(@Autowired))`](https://projectlombok.org/api/lombok/RequiredArgsConstructor.html).

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/data_migration/BlogPostUuidShorteningConvertingScheduler.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+@Component
+@Slf4j
+public class BlogPostUuidShorteningConvertingScheduler {
+
+    private static final int BATCH_SIZE = 250;
+
+    private final BlogPostUidShortener uuidShortener;
+
+    @Autowired
+    public BlogPostUuidShorteningConvertingScheduler(BlogPostUidShortener uuidShortener) {
```
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/data_migration/BlogPostUuidShorteningConvertingScheduler.java]: This handcrafted constructor can be replaced with class level annotation [`@RequiredArgsConstructor(onConstructor = @__(@Autowired))`](https://projectlombok.org/api/lombok/RequiredArgsConstructor.html).

**Comment by matek2305** [comment]
Why is this done with scheduler? Shouldn't it be some script that will runned during db schema migration and then app will generate valid uids. With this solution scheduler will be runned f.e. 10 times every 1 minute while it shorten all uids and then there will be no work for it, am I missing something?

**Comment by goostleek** [comment]
I'm wondering if we really need UUID shortening logic at all. The code responsible for the shortening the UUIDs will be executed only once. Once it's done the code is not needed anymore because we already have short UUIDs. The conversion should be done by some script. WDYT?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java [hunk 4]
```diff
@@ -95,4 +95,73 @@ public String getApprovalState() {
     public boolean isGoingInNewsletter(final LocalDateTime lastPublicationDate) {
         return publishedDate.isAfter(lastPublicationDate);
     }
+
+    public static BlogPostBuilder builder() {
+        return new BlogPostBuilder();
+    }
+
+    public static class BlogPostBuilder {
```
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java]: Is this builder really required? It's not used anywhere.

**Comment by tdziurko** [comment]
@goostleek, @matek2305 I planned to keep this scheduler for a while so other developers could have this migration applied to their databases to keep their state consistent. But maybe this drawback is a not a drawback at all if you suggest not putting it into the repository.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java (hunk 4 above)
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java]: It's present in NewBlogPostStoringActor

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java [hunk 5]
```diff
@@ -95,4 +95,73 @@ public String getApprovalState() {
     public boolean isGoingInNewsletter(final LocalDateTime lastPublicationDate) {
         return publishedDate.isAfter(lastPublicationDate);
     }
+
+    public static BlogPostBuilder builder() {
```
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java]: Is this builder really required? It's not used anywhere so it can be remove IMHO.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java [hunk 6]
```diff
@@ -31,14 +31,14 @@
...
-@Builder
 @EntityListeners(BlogPostEntityListener.class)
 public class BlogPost {
 
     public static final int MAX_DESCRIPTION_LENGTH = 4096;
+    public static final int UID_LENGTH = 7;
 
     @Column(name = "UID", unique = true, nullable = false)
-    private final String uid = UUID.randomUUID().toString();
+    private String uid = generateRandomUid();
```
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java]: The field can still be final. Why did you remove the modifier?

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java (hunk 5 above)
Comment (tdziurko) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java]: @Builder populates uid with null value when not set explicitly so we have to do it manually.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java [hunk 7]
```diff
@@ -95,4 +95,73 @@ public String getApprovalState() {
...
+        public BlogPost.BlogPostBuilder blog(Blog blog) {
+            this.blog = blog;
+            return this;
+        }
+
+        public BlogPost build() {
+            return new BlogPost(id, title, description, url, publishedDate, approved, blog);
+        }
+
+        public String toString() {
```
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java]: Custom 'toString()' implementation can be replaced by '@ToString' annotation.

File: src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java (hunk 5 above)
Comment (goostleek) [inline src/main/java/pl/tomaszdziurko/jvm_bloggers/blog_posts/domain/BlogPost.java]: It won't set it to `null` anymore after you made the `uid` field final. You can safely replace this custom builder implementation back to `@Builder`.
